File: server/src/emberline/context/ring.py
"""Cross-file context: a ring buffer of chunks from other open files.

Modelled on llama.vim's ring buffer, which ranks chunks by plain line-set overlap
rather than embeddings. That choice is deliberate and worth preserving: ranking
runs on every keystroke, so it has to be effectively free. Embeddings are reserved
for the accepted-example store, where lookups are rarer and the text is short.

Chunks land in ``/infill``'s ``input_extra``, which llama.cpp inserts ahead of the
FIM prefix.
"""

from __future__ import annotations

import logging
import re
from collections import OrderedDict
from dataclasses import dataclass
from pathlib import Path

log = logging.getLogger(__name__)

_TOKEN_RE = re.compile(r"[A-Za-z_][A-Za-z0-9_]{2,}")
_MAX_FILE_BYTES = 512 * 1024
"""Skip anything bigger; it is a bundle, a lockfile, or generated."""


def _tokens(text: str) -> set[str]:
    return set(_TOKEN_RE.findall(text))


def _similarity(a: set[str], b: set[str]) -> float:
    """Jaccard over identifier-ish tokens."""
    if not a or not b:
        return 0.0
    inter = len(a & b)
    if inter == 0:
        return 0.0
    return inter / len(a | b)


@dataclass(slots=True)
class Chunk:
    filename: str
    text: str
    tokens: set[str]


class RingContext:
    def __init__(self, *, max_chunks: int = 16, chunk_lines: int = 64) -> None:
        self._max_chunks = max_chunks
        self._chunk_lines = chunk_lines
        # path -> (mtime_ns, chunks). Keeps us off the disk on every keystroke.
        self._files: OrderedDict[str, tuple[int, list[Chunk]]] = OrderedDict()
        self._max_files = 32

    def _chunks_for(self, path: str) -> list[Chunk]:
        p = Path(path)
        try:
            stat = p.stat()
        except OSError:
            self._files.pop(path, None)
            return []
        if stat.st_size > _MAX_FILE_BYTES:
            return []

        cached = self._files.get(path)
        if cached is not None and cached[0] == stat.st_mtime_ns:
            self._files.move_to_end(path)
            return cached[1]

        try:
            text = p.read_text(encoding="utf-8", errors="strict")
        except (OSError, UnicodeDecodeError):
            # Binary or unreadable; remember nothing rather than retrying each keystroke.
            self._files[path] = (stat.st_mtime_ns, [])
            return []

        lines = text.splitlines()
        chunks: list[Chunk] = []
        for start in range(0, len(lines), self._chunk_lines):
            body = "\n".join(lines[start : start + self._chunk_lines])
            if not body.strip():
                continue
            chunks.append(Chunk(filename=p.name, text=body, tokens=_tokens(body)))

        self._files[path] = (stat.st_mtime_ns, chunks)
        self._files.move_to_end(path)
        while len(self._files) > self._max_files:
            self._files.popitem(last=False)
        return chunks

    def build(
        self,
        *,
        prefix: str,
        current_path: str,
        open_paths: list[str],
    ) -> list[dict[str, str]]:
        """Rank chunks from other open files against the cursor context."""
        query = _tokens(prefix[-4000:])
        if not query:
            return []

        scored: list[tuple[float, Chunk]] = []
        for path in open_paths:
            if path == current_path:
                continue
            for chunk in self._chunks_for(path):
                score = _similarity(query, chunk.tokens)
                if score > 0.0:
                    scored.append((score, chunk))

        scored.sort(key=lambda pair: pair[0], reverse=True)
        top = scored[: self._max_chunks]
        return [{"filename": c.filename, "text": c.text} for _, c in top]

File: extension/src/engine/ring.ts
/**
 * Cross-file context: a ring buffer of chunks from other open files.
 *
 * Modelled on llama.vim's ring buffer, which ranks chunks by plain line-set overlap
 * rather than embeddings. That choice is deliberate and worth preserving: ranking
 * runs on every keystroke, so it has to be effectively free.
 *
 * Chunks land in `/infill`'s `input_extra`, which llama.cpp inserts ahead of the
 * FIM prefix.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';

import { similarity, tokens } from './tokens';
import type { ExtraChunk } from './types';

/** Skip anything bigger; it is a bundle, a lockfile, or generated. */
const MAX_FILE_BYTES = 512 * 1024;

/**
 * Rejects invalid UTF-8 instead of substituting U+FFFD.
 *
 * This is load-bearing, and it is the one place where the obvious port is wrong:
 * `fs.readFileSync(p, 'utf8')` never throws on a binary file, it silently fills
 * the text with replacement characters. Python's `read_text(errors="strict")`
 * raised, which is how binaries got skipped -- without `fatal` they would instead
 * be chunked into garbage and fed to the model as cross-file context.
 */
const DECODER = new TextDecoder('utf-8', { fatal: true });

interface Chunk {
	filename: string;
	text: string;
	tokens: Set<string>;
}

interface CachedFile {
	mtimeNs: bigint;
	chunks: Chunk[];
}

export class RingContext {
	private readonly maxChunks: number;
	private readonly chunkLines: number;
	// path -> chunks. Insertion-ordered, so the first key is the least recently
	// used. Keeps us off the disk on every keystroke.
	private readonly files = new Map<string, CachedFile>();
	private readonly maxFiles = 32;

	constructor(opts: { maxChunks?: number; chunkLines?: number } = {}) {
		this.maxChunks = opts.maxChunks ?? 16;
		this.chunkLines = opts.chunkLines ?? 64;
	}

	private chunksFor(p: string): Chunk[] {
		let stat: fs.BigIntStats;
		try {
			// bigint for exact mtime: mtimeMs is a float and loses the resolution
			// this cache check depends on.
			stat = fs.statSync(p, { bigint: true });
		} catch {
			this.files.delete(p);
			return [];
		}
		if (stat.size > BigInt(MAX_FILE_BYTES)) {
			return [];
		}

		const cached = this.files.get(p);
		if (cached !== undefined && cached.mtimeNs === stat.mtimeNs) {
			this.files.delete(p);
			this.files.set(p, cached);
			return cached.chunks;
		}

		let text: string;
		try {
			text = DECODER.decode(fs.readFileSync(p));
		} catch {
			// Binary or unreadable; remember nothing rather than retrying each keystroke.
			this.remember(p, { mtimeNs: stat.mtimeNs, chunks: [] });
			return [];
		}

		const filename = path.basename(p);
		const lines = text.split(/\r\n|\r|\n/);
		const chunks: Chunk[] = [];
		for (let start = 0; start < lines.length; start += this.chunkLines) {
			const body = lines.slice(start, start + this.chunkLines).join('\n');
			if (!body.trim()) {
				continue;
			}
			chunks.push({ filename, text: body, tokens: tokens(body) });
		}

		this.remember(p, { mtimeNs: stat.mtimeNs, chunks });
		return chunks;
	}

	private remember(p: string, entry: CachedFile): void {
		this.files.delete(p);
		this.files.set(p, entry);
		while (this.files.size > this.maxFiles) {
			const oldest = this.files.keys().next();
			if (oldest.done) {
				break;
			}
			this.files.delete(oldest.value);
		}
	}

	/** Rank chunks from other open files against the cursor context. */
	build(args: { prefix: string; currentPath: string; openPaths: string[] }): ExtraChunk[] {
		const query = tokens(args.prefix.slice(-4000));
		if (query.size === 0) {
			return [];
		}

		const scored: Array<{ score: number; chunk: Chunk }> = [];
		for (const p of args.openPaths) {
			if (p === args.currentPath) {
				continue;
			}
			for (const chunk of this.chunksFor(p)) {
				const score = similarity(query, chunk.tokens);
				if (score > 0) {
					scored.push({ score, chunk });
				}
			}
		}

		// Array.prototype.sort is stable (ES2019+), matching Python's sort: ties
		// keep open_paths order rather than shuffling between keystrokes.
		scored.sort((a, b) => b.score - a.score);
		return scored
			.slice(0, this.maxChunks)
			.map(({ chunk }) => ({ filename: chunk.filename, text: chunk.text }));
	}
}
